File: verifyinganaliendictionary.js
//Objective is to see whether words in a list are in valid lexicographic order
//based on an alien dictionary.

let words = ["hello","leetcode"]
let order = "hlabcdefgijkmnopqrstuvwxyz"


//O(n) solution that uses a hashmap to store the indexes of each character
//in the order.

let map = {}

//Map the order of the alien dictionary
for (let i = 0; i < order.length; i++) {
    map[order.charAt(i)] = i
}

for (let i = 1; i < words.length; i++) {
    let word1 = words[i - 1]
    let word2 = words[i]

    if (map[word1[0]] > map[word2[0]]) {
        return false
    } else if (map[word1[0]] == map[word2[0]]) {
        let index = 1

        //Skip all of the same characters
        while (word1[index] == word2[index]) {
            index++
        }

        //If word1 and word2 share the same characters but
        //word1 is longer, return false
        if (!word2[index]) {
            return false
        }

        if (map[word1[index]] > map[word2[index]]) {
            return false
        }
    }
}

return true

File: verifyinganaliendictionary.java
//Java Solution

class Solution {
    public boolean isAlienSorted(String[] words, String order) {
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < order.length(); i++) {
            map.put(order.charAt(i), i);
        }
        
        for (int i = 1; i < words.length; i++) {
            String word1 = words[i - 1];
            String word2 = words[i];
            
            if (map.get(word1.charAt(0)) > map.get(word2.charAt(0))) {
                return false;
            } else if (word1.charAt(0) == word2.charAt(0)) {
                int p1 = 1;
                
                while (p1 < word1.length() && p1 < word2.length() && word1.charAt(p1) == word2.charAt(p1)) {
                    p1++;
                }
                
                if (p1 == word2.length()) {
                    return false;
                }
                
                if (p1 < word1.length() && p1 < word2.length() && map.get(word1.charAt(p1)) > map.get(word2.charAt(p1))) {
                    return false;
                }
            }
        }
        
        return true;
    }
}